.changeset/slow-items-heal.md

```diff
@@ -0,0 +1,36 @@
+---
+"astro": minor
+---
+
+Adds experimental JSON Schema support for content collections of with `type: 'data'`.
+
+To enable this feature, add the experimental flag:
+
+```diff
+import { defineConfig } from 'astro/config';
+
+export default defineConfig({
+	experimental: {
++		contentCollectionJsonSchema: true
+	}
+});
+```
+
+To be able to use the schemas, you need to adapt your zod definition and reference it inside the json manually:
+
+```diff
+const test = defineCollection({
+	type: 'data',
+	schema: z.object({
++		"$schema": z.string().optional(),
+    test: z.string()
+	}),
+});
+```
+
+```diff
+{
++  "$schema": "../../../.astro/schemas/collections/test.json",
```

Comment:
This is valid and straight-forward, but you can also adjust your workspace settings in VSCode to avoid needing to set `$schema` on every entry. Might be worth mentioning!

```
"json.schemas": [
  {
    "fileMatch": [
      "/src/content/test/**"
    ],
    "url": "../../../.astro/schemas/collections/test.json"
  }
]
```

It would be nice to link directly to the [VS Code JSON docs](https://code.visualstudio.com/docs/languages/json#_json-schemas-and-settings)